docs/tutorial/2_MarketMaker.md

```diff
@@ -0,0 +1,230 @@
+
+I want to make sure that I don't get less than the value that was promised by `quote`. So I fill 471765614363247798186 (`quote`'s answer') in `minReturn` argument. After a while, checking the [transaction](https://ropsten.etherscan.io/tx/0x6c711ea96f507a8e71330fb5a3f4b675c81eda3083a2b2b707217802f199e673) I see that this is exactly the amount of tokens I got. Looks solid.
+
+Now let's try to exchange another 1000 CLN for the same exchange rate. I know this is impossible but for once I want to show you a failed transaction. Actually Ethereum doesn't give a reason why the [transaction failed](https://ropsten.etherscan.io/tx/0x1bc68cacb5bf67dcb0d3c0ee8eedf124171aadeaaa52f5107f5806649aa31a05), so you should trust me on that one. But when I run again `quote` for 1000 CLN I'm getting a `returnAmount` of ~366 CC, so this makes sense.
+
+
+## The Math Behind
+
+The last thing I want to cover is the formula itself. If you want to have a deep understanding of the math I advise you to read the whitepaper's [appendix](https://cln.network/pdf/cln_whitepaper.pdf).
+
+### The Ellipse
+
+Theoretically, the Market Maker can work with any pair of currencies, but for our use case the first currency is always CLN and the second one is the Community Currency. As I explained before, there are multiple Community Currencies with a dedicated Market Maker for each one. But each CC uses CLN as backup currency, so through CLN you can exchange CC1 for CC2 in two hops.
+
+There is four variables that affect the exchange rate:
+- S1 - total supply of the first currency, the CLN. Namely, the number of CLN tokens issued, we can consider it as a constant.
+- S2 - total supply of the second currency, the CC. It's defined when the currency is created, and is also a constant.
+- R1 - the reservoir of the first currency held by the Market Maker. Initially it's zero cause no CLN is inserted.
+- R2 - the reservoir of the second currency held by the Market Maker. Initially all the CC's total supply is in the reservoir, so R2 = S2.
+
+These four variables should apply to the formula:
+
+![ellipse](../assets/formulas/ellipse.gif)
+
+As I said, S1 and S2 are constants. So if denoted `x = S1 - R1` and `y = S2 - R2`, we get an [ellipse equation](https://en.wikipedia.org/wiki/Ellipse#Equation), that's why the contract is called Ellipse Market Maker :open_mouth: !
+
+Let's check that the formula holds at least for a couple of cases:
+
+#### Issuance
+After the issuance we know that:
+
+```
+R1 = 0
+R2 = S2
+```
+
+Let's substitute this to the formula
+```
+=> (S1 / S1) + (S2 - S2 / S2) = 1
+
+=> 1 = 1
+```
+Q.E.D. :white_check_mark:
+
+#### 1000 CLN inserted
+
+Let's take parameters from the issuance in tutorial's first part. The constants are:
```

Comment:
Let's take parameters from the issuance in tutorial's first part --> Let's take the parameters from the issuance in the first part of the tutorial